File: src/ruby/tormenta20/models/escudo.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing an Escudo (Shield) in Tormenta20.
    #
    # Shields provide additional defense bonuses but may impose
    # armor penalties on certain skill checks.
    #
    # @example Find all shields
    #   Tormenta20.escudos.all
    #
    # @example Find shield with highest defense
    #   Tormenta20.escudos.order(defense_bonus: :desc).first
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "escudo_leve")
    # @!attribute [rw] name
    #   @return [String] Shield name in Portuguese
    # @!attribute [rw] price
    #   @return [String] Price in Tormenta currency
    # @!attribute [rw] defense_bonus
    #   @return [Integer] Defense bonus provided by the shield
    # @!attribute [rw] armor_penalty
    #   @return [Integer] Armor penalty for skill checks
    # @!attribute [rw] weight
    #   @return [String] Weight in slots/spaces
    # @!attribute [rw] properties
    #   @return [Array<String>] Special shield properties
    # @!attribute [rw] description
    #   @return [String, nil] Optional description
    class Escudo < Base
      self.table_name = "escudos"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :defense_bonus, presence: true

      # Convert the shield to a hash representation.
      #
      # @return [Hash] Hash with all shield attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          price: price,
          defense_bonus: defense_bonus,
          armor_penalty: armor_penalty,
          weight: weight,
          properties: properties,
          description: description
        }.compact
      end
    end
  end
end

File: src/ts/models/escudo.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "escudos"

/**
 * An Escudo (shield) in Tormenta20.
 *
 * @example
 * ```ts
 * Escudo.all()
 * Escudo.find("escudo_leve")
 * ```
 */
export class Escudo extends BaseModel {
  /** Price in tibares. */
  get price(): number { return (this._row.price as number) ?? 0 }
  /** Bonus added to Defense. */
  get defenseBonus(): number { return this._row.defense_bonus as number }
  /** Penalty applied to Agility-based skills and attack rolls. */
  get armorPenalty(): number { return (this._row.armor_penalty as number) ?? 0 }
  /** Weight in spaces (espaços). */
  get weight(): number { return (this._row.weight as number) ?? 0 }
  /** JSON array of special property objects. */
  get properties(): unknown[] { return BaseModel.json(this._row.properties, []) }
  get description(): string | null { return (this._row.description as string) ?? null }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, price: this.price, defense_bonus: this.defenseBonus,
      armor_penalty: this.armorPenalty, weight: this.weight,
      properties: this.properties, description: this.description,
    }
  }

  static query(): Query<Escudo> { return BaseModel.makeQuery(TABLE, Escudo) }
  static all(): Escudo[] { return Escudo.query().all() }
  static find(id: string): Escudo | null { return Escudo.query().find(id) }
  static first(): Escudo | null { return Escudo.query().first() }
  static last(): Escudo | null { return Escudo.query().last() }
  static count(): number { return Escudo.query().count() }
}
